lib/kuzzle/internalIndexHandler.js

```diff
@@ -0,0 +1,241 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2020 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const crypto = require('crypto');
+
+const Bluebird = require('bluebird');
+
+const debug = require('../util/debug')('kuzzle:bootstrap:internalIndex');
+const Store = require('../core/shared/store');
+const Mutex = require('../util/mutex');
+const ApiKey = require('../model/storage/apiKey');
+const scopeEnum = require('../core/storage/storeScopeEnum');
+const kerror = require('../kerror');
+
+const securitiesBootstrap = {
+  profiles: {
+    admin: {
+      policies: [ { roleId: 'admin' } ],
+      rateLimit: 0,
+    },
+    anonymous: {
+      policies: [ { roleId: 'anonymous' } ],
+    },
+    default: {
+      policies: [ { roleId: 'default' } ],
+    }
+  },
+  roles: {
+    admin: {
+      controllers: {
+        '*': {
+          actions: {
+            '*': true,
+          },
+        },
+      },
+    },
+    anonymous: {
+      controllers: {
+        '*': {
+          actions: {
+            '*': true,
+          },
+        },
+      },
+    },
+    default: {
+      controllers: {
+        '*': {
+          actions: {
+            '*': true,
+          },
+        },
+      },
+    },
+  },
+};
+
+const dataModelVersion = '2.0.0';
+
+class InternalIndexHandler extends Store {
+  constructor (kuzzle) {
+    super(
+      kuzzle,
+      kuzzle.config.services.storageEngine.internalIndex.name,
+      scopeEnum.PRIVATE);
+
+    this.timeout = kuzzle.config.services.internalIndex.bootstrapLockTimeout;
+    this.config = kuzzle.config.services.storageEngine.internalIndex;
+
+    // IDs for config documents
+    this._BOOTSTRAP_DONE_ID = `${this.index}.done`;
+    this._DATAMODEL_VERSION_ID = 'internalIndex.dataModelVersion';
+    this._JWT_SECRET_ID = 'security.jwt.secret';
+  }
+
+  /**
+   * @returns {Promise}
+   */
+  async init () {
+    await super.init(this.config.collections);
+
+    const mutex = new Mutex(this.kuzzle, 'InternalIndexBootstrap', {
+      timeout: -1,
+      ttl: 30000,
+    });
+
+    await mutex.lock();
```

Comment:
Can't we check if the internal index is bootstrapped before trying to lock the mutex? 

Actually each node will have to wait until the mutex is free to check if the internal index exists or not.

Reply:
We can't do that, because the "bootstrap done" document could be created between the check and the mutex lock.